src/charm.py

```diff
@@ -0,0 +1,257 @@
+        c = ' && '.join(
+            [
+                'bash /etc/s6-overlay/s6-rc.d/init-calibre-web-config/run',
+                # with bash because the run script shebang depends on s6
+                'export CALIBRE_DBPATH=/config',
+                'cd /app/calibre-web',
+                'python3 /app/calibre-web/cps.py',
+            ]
+        )
+        command = f"bash -c '{c}'"
+        return {
+            'summary': f'{SERVICE_NAME} layer',
+            'description': f'pebble config layer for {SERVICE_NAME}',
+            'services': {
+                SERVICE_NAME: {
+                    'override': 'replace',
+                    'summary': 'calibre-web',
+                    'command': command,
+                    'startup': 'enabled',
+                    'environment': {
+                        'PUID': '1000',  # copied from example docker run
+                        'PGID': '1000',  # copied from example docker run
+                        'TZ': 'Etc/UTC',  # copied from example docker run
+                    },
+                }
+            },
+        }
+
+    def _push_library_to_storage(self) -> None:
+        """Push user provided or default calibre-library resource to storage."""
+        library_write_behaviour, _ = self._get_library_write_behaviour()
+        if library_write_behaviour is None:
+            return
+        container = self.framework.model.unit.containers[CONTAINER_NAME]
+        if contents := container.list_files('/books/'):
+            msg_prefix = f'_push_library_to_storage: {library_write_behaviour=}: '
+            match library_write_behaviour:
+                case 'skip':
+                    logger.debug(msg_prefix + 'returning')
+                    return
+                case 'clean':
+                    logger.debug(msg_prefix + 'cleaning ...')
+                    for fileinfo in contents:
+                        container.remove_path(fileinfo.path, recursive=True)
+        logger.debug(f'_push_library_to_storage: {container=}')
+        library_path = Path(self.framework.model.resources.fetch('calibre-library'))
+        library = library_path.read_bytes()
+        # if the default library 'resource' was the starter library.zip
+        # instead of the empty empty.zip, this logic could be removed
+        if len(library):  # user provided library
+            logger.debug(f'_push_library_to_storage: {library_path=}')
+            self._push_and_extract_library(container, library)
+            return
+        # else: use default library
+        library_path = Path('.') / 'library.zip'
+        library = library_path.read_bytes()
+        assert library
+        logger.debug(f'_push_library_to_storage: {library_path=}')
+        self._push_and_extract_library(container, library)
+
+    def _get_library_write_behaviour(
+        self,
+    ) -> tuple[LibraryWriteBehaviour, ops.ActiveStatus] | tuple[None, ops.BlockedStatus]:
```

Comment:
It is a bit ugly looking! I borrowed from golang's idiom for error handling here in my attempt to keep the status-evaluating and value-returning code in one place -- it's called both in the code that needs the config value (or to abort if it's invalid), and in the `collect-status` observer to add the status.

I originally threw an exception, but that made it tricky to get the bad config value for the `blocked` message. I mean you could attach it to the `Exception`, but that feels a bit nasty.

An alternative with a nicer return value annotation would be `LibraryWriteBehaviour | ops.BlockedStatus`, but it would make the code at the two calling sites (imo) a bit uglier ...

The planned changes you mention would definitely help with this case, but it seems like keeping your status evaluations dry / single source of truth is still a bit tricky, and I'd like to learn or figure out the best idioms there.

Reply:
I use this in three places currently:
```python
def _on_collect_status(self, event: ops.CollectStatusEvent) -> None:
    ...
    _, status = self._get_library_write_behaviour()
    event.add_status(status)

def _on_library_write(self, event: ops.ActionEvent) -> None:
    library_write_behaviour, status = self._get_library_write_behaviour()
    if library_write_behaviour is None:
        event.fail(status.message)
        return
    self._push_library_to_storage()  # also triggered during initial setup
    event.set_results({LIBRARY_WRITE_CONFIG: library_write_behaviour})

def _push_library_to_storage(self) -> None:
    library_write_behaviour, _ = self._get_library_write_behaviour()
    if library_write_behaviour is None:
        return
    ...
    match library_write_behaviour:
        ...
```
I guess I'm wondering if this is a good pattern for handling this kind of thing? Is there a more preferred idiom currently? Or is the fact that I'm reaching for a solution like this a sign that I've taken a wrong turn somewhere?

Reply:
> It is a bit ugly looking! I borrowed from golang's idiom for error handling here

Indeed, it's ugly exactly like Go :stuck_out_tongue: .

> I originally threw an exception, but that made it tricky to get the bad config value for the `blocked` message. I mean you could attach it to the `Exception`, but that feels a bit nasty.

This doesn't seem nasty?

```python
def get_config(self):
    raise ValueError(f"your {value} is bad, dummy")

def use_config(self):
    try:
        config = get_config()
    except ValueError as e:
        event.fail(str(e))  # or event.fail(e.args[0])
```

> An alternative with a nicer return value annotation would be `LibraryWriteBehaviour | ops.BlockedStatus`, but it would make the code at the two calling sites (imo) a bit uglier ...

I'm definitely -1 on having a return type that's a mix of two entirely different types (except for `None` and something).

> The planned changes you mention would definitely help with this case, but it seems like keeping your status evaluations dry / single source of truth is still a bit tricky, and I'd like to learn or figure out the best idioms there.

I would do something like this:

```python
def _on_collect_status(self, event: ops.CollectStatusEvent) -> None:
    ...
    try:
        self._get_library_write_behaviour()  # Ignore return value, we just want to validate.
    except ValueError as e:
        event.add_status(ops.BlockedStatus(str(e))

def _on_library_write(self, event: ops.ActionEvent) -> None:
    try:
        library_write_behaviour self._get_library_write_behaviour()
    except ValueError as e:
        event.fail(str(e))
        return
    self._push_library_to_storage()  # also triggered during initial setup
    event.set_results({LIBRARY_WRITE_CONFIG: library_write_behaviour})

def _push_library_to_storage(self) -> None:
    try:
        library_write_behaviour = self._get_library_write_behaviour()
    except ValueError:
        return
    ...
    match library_write_behaviour:
        ...
```

Reply:
Your proposed solution is probably idea in this case.

But while we're here, how would you handle things if a method could result in (say) either a `blocked` or `waiting` status? Would the ugly golang idiom be better here?
```python
def f(self) -> tuple[Value, None] | tuple[None, BadStatus]:
```
Or maybe a mapping of (custom?) exception types to statuses, either generally or just in a series of `except` blocks in `_on_collect_status`?

Reply:
I've seen people do something like:

```
class InvalidConfigError(Exception);
    def __init__(self, message, status):
        self.message = message
        self.status = status

class MyCharm(...):
    ...
    def validate(self):
        if something:
            raise InvalidConfig("message for logging", ops.BlockedStatus("blah"))
```

There are other people (maybe all of them!) in our team that like Go more than me, so might be more inclined to the (value, err) style, although it still feels unpythonic to me.

I think in most cases you actually only expect to have a specific status - if you're getting config, then I'd expect that you end up in blocked, if you are getting relation data I'd expect you end up in Waiting, etc. So I think you probably do end up just raising different exception types (probably still custom ones) depending on the situation.

Reply:
Thanks for the explanation, Tony, that's really helpful. I think of myself as more of a python guy, but transferring information around on exceptions does feel a little bit off to me compared to returning the information directly ...

Here's a wild idea for your entertainment - make the status classes all inherit from `Exception` (via some lineage) so that you can do:
```python
try:
    something()
except (BlockedStatus, MaintenanceStatus, WaitingStatus) as e:
    event.add_status(e)
```

How to interpret `raise ops.ActiveStatus()` is left as an exercise for the reader.

Reply:
Ha, I could see all but active as exceptions, where somewhere in the framework they are caught and set the status and provide a way to jump out of the code. In another timeline ...